include/onnxruntime/core/framework/int4.h

```diff
@@ -0,0 +1,143 @@
+// Copyright (c) Microsoft Corporation. All rights reserved.
+// Licensed under the MIT License.
+
+#pragma once
+
+#include <cassert>
+#include <type_traits>
+#include "core/common/common.h"
+#include "core/common/gsl.h"
+
+namespace onnxruntime {
+
+template <bool Signed>
+struct Int4Traits;
+
+template <>
+struct Int4Traits<true> {
+  using UnpackedType = int8_t;
+  static constexpr int8_t min_val = -8;
+  static constexpr int8_t max_val = 7;
+};
+
+template <>
+struct Int4Traits<false> {
+  using UnpackedType = uint8_t;
+  static constexpr uint8_t min_val = 0;
+  static constexpr uint8_t max_val = 15;
+};
+
+/// <summary>
+/// Stores 2 packed 4-bit elements in 1 byte.
+/// </summary>
+/// <typeparam name="Signed">Set to true if signed int4, or false if unsigned uint4.</typeparam>
+template <bool Signed>
+struct Int4x2Base {
+  using UnpackedType = typename Int4Traits<Signed>::UnpackedType;
+  static constexpr UnpackedType min_val = Int4Traits<Signed>::min_val;
+  static constexpr UnpackedType max_val = Int4Traits<Signed>::max_val;
+
+  uint8_t bits_{};
+
+  Int4x2Base() = default;
+
+  explicit Int4x2Base(uint8_t bits) {
+    bits_ = bits;
+  }
+
+  Int4x2Base(UnpackedType val0, UnpackedType val1) {
+    bits_ = static_cast<uint8_t>(((val1 & 0xF) << 4) | (val0 & 0xF));
+  }
+
+  static inline int8_t SignExtendLower4Bits(uint8_t bits) {
+    // Sign-extend lower 4-bits by left shifting and then doing an arithmetic right shift.
+    constexpr uint8_t Shift = (sizeof(int32_t) * 8) - 4;
+    return static_cast<int8_t>((static_cast<int32_t>(bits) << Shift) >> Shift);
+  }
+
+  inline UnpackedType GetElem0() const {
+    if constexpr (Signed) {
+      return SignExtendLower4Bits(bits_);
+    } else {
+      return static_cast<UnpackedType>(bits_ & 0xF);
+    }
+  }
+
+  inline UnpackedType GetElem1() const {
+    const uint8_t val = static_cast<uint8_t>((bits_ >> 4) & 0xF);
+
+    if constexpr (Signed) {
+      return SignExtendLower4Bits(val);
+    } else {
+      return val;
+    }
+  }
+
+  inline UnpackedType GetElem(size_t index) const {
```

Comment:
To avoid having to handle an out-of-bounds index, consider making the parameter an enum that can only be 0 or 1.

```c++
inline UnpackedType GetElem(Int4x2Element which) const {
```